TestProjects/ShaderGraph/Assets/CommonAssets/Editor/ShaderGraphTestRenderer.cs

```diff
@@ -0,0 +1,246 @@
+using System;
+using System.IO;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEditor;
+using UnityEngine.Experimental.Rendering;
+using UnityEngine.Rendering;
+using UnityEditor.ShaderGraph;
+using UnityEditor.ShaderGraph.Drawing;
+using NUnit.Framework;
+
+public class ShaderGraphTestRenderer
+{
+    internal const int defaultResolution = 128;
+
+    internal PreviewSceneResources previewScene = new PreviewSceneResources();
+
+    internal delegate void SetupMaterialDelegate(Material m);
+
+    internal int wrongImageCount;
+    internal int mostWrongPixels;
+    internal string mostWrongString;
+    internal void ResetTestReporting()
+    {
+        wrongImageCount = 0;
+        mostWrongPixels = 0;
+        mostWrongString = string.Empty;
+    }
+    internal void ReportTests()
+    {
+        if (wrongImageCount > 0)
+        {
+            Assert.That(false, $"{wrongImageCount} images failed, worst was: {mostWrongString}");
+        }
+    }
+
+    internal GraphData LoadGraph(string graphPath)
+    {
+        List<PropertyCollector.TextureInfo> lti;
+        var assetCollection = new AssetCollection();
+        ShaderGraphImporter.GetShaderText(graphPath, out lti, assetCollection, out var graph);
+        Assert.NotNull(graph, $"Invalid graph data found for {graphPath}");
+        graph.OnEnable();
+        graph.ValidateGraph();
+        return graph;
+    }
+
+    // we apply a transform to the test setup, so that the transform matrices are non-trivial
+    internal Vector3 testPosition = new Vector3(0.24699998f, 0.51900005f, 0.328999996f);
+    internal Quaternion testRotation = new Quaternion(-0.164710045f, -0.0826543793f, -0.220811233f, 0.957748055f);
+
+    internal int RunNodeTest(GraphData graph, string filePrefix, SetupMaterialDelegate setupMaterial = null, Color32? expectedColor = null, int expectedIncorrectPixels = 0)
+    {
+        RenderTextureDescriptor descriptor = new RenderTextureDescriptor(defaultResolution, defaultResolution, GraphicsFormat.R8G8B8A8_SRGB, depthBufferBits: 32);
+        var target = RenderTexture.GetTemporary(descriptor);
+
+        // use a non-standard transform, so that view, object, etc. transforms are non trivial
+        RenderQuadPreview(graph, target, testPosition, testRotation, setupMaterial, Mode.DIFF, useSRP: true);
+
+        int incorrectPixels = CountPixelsNotEqual(target, expectedColor ?? new Color32(0, 255, 0, 255), false);
```

Comment:
Why does the default expected color have green? This seems odd, perhaps a small comment?

Reply:
added a comment in latest